testsuite/model/src/test/java/org/keycloak/testsuite/model/session/HotRodUserSessionClientSessionRelationshipTest.java

```diff
@@ -0,0 +1,144 @@
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserSessionModel;
+import org.keycloak.models.UserSessionProvider;
+import org.keycloak.models.map.storage.ModelEntityUtil;
+import org.keycloak.models.map.storage.hotRod.connections.DefaultHotRodConnectionProviderFactory;
+import org.keycloak.models.map.storage.hotRod.connections.HotRodConnectionProvider;
+import org.keycloak.models.map.storage.hotRod.userSession.HotRodUserSessionEntity;
+import org.keycloak.testsuite.model.KeycloakModelTest;
+import org.keycloak.testsuite.model.RequireProvider;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.aMapWithSize;
+import static org.hamcrest.Matchers.anEmptyMap;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+
+@RequireProvider(UserSessionProvider.class)
+@RequireProvider(value = HotRodConnectionProvider.class, only = DefaultHotRodConnectionProviderFactory.PROVIDER_ID)
+public class HotRodUserSessionClientSessionRelationshipTest extends KeycloakModelTest {
+
+    private String realmId;
+    private String CLIENT0_CLIENT_ID = "client0";
+
+    @Override
+    public void createEnvironment(KeycloakSession s) {
+        RealmModel realm = s.realms().createRealm("test");
+        realm.setDefaultRole(s.roles().addRealmRole(realm, Constants.DEFAULT_ROLES_ROLE_PREFIX + "-" + realm.getName()));
+        realm.setSsoSessionIdleTimeout(1800);
+        realm.setSsoSessionMaxLifespan(36000);
+        this.realmId = realm.getId();
+        s.clients().addClient(realm, CLIENT0_CLIENT_ID);
+
+        s.users().addUser(realm, "user1").setEmail("user1@localhost");
+    }
+
+    @Override
+    public void cleanEnvironment(KeycloakSession s) {
+        if (realmId != null) {
+            s.realms().removeRealm(realmId);
+        }
+    }
+
+    @Test
+    public void testClientSessionAreRemovedOnUserSessionRemoval() {
+        AtomicReference<String> uSessionId = new AtomicReference<>();
+        AtomicReference<String> cSessionId = new AtomicReference<>();
+        prepareSessions(uSessionId, cSessionId);
+
+        withRealm(realmId, (session, realm) -> {
+            UserSessionModel uSession = session.sessions().getUserSession(realm, uSessionId.get());
+            session.sessions().removeUserSession(realm, uSession);
+            return null;
+        });
+
+        assertCacheContains(remoteCache -> assertThat(remoteCache, anEmptyMap()));
+    }
+
+    @Test
+    public void testSessionsAreRemovedOnUserRemoval() {
+        AtomicReference<String> uSessionId = new AtomicReference<>();
+        AtomicReference<String> cSessionId = new AtomicReference<>();
+        prepareSessions(uSessionId, cSessionId);
+
+        withRealm(realmId, (session, realm) -> {
+            session.clients().removeClient(realm, CLIENT0_CLIENT_ID);
```

Comment:
It is also necessary to remove the client here?

Reply:
No, it is not. It is a relict from a similar test I created for clients. Then I realised we are not removing client sessions on client removal. We are just checking whether the client exists on querying client sessions.